static/js/components/ShiftPage.jsx

```diff
@@ -28,30 +29,87 @@ function datestringToDate(shiftList) {
   return shiftList;
 }
 
-const ShiftPage = () => {
+const ShiftPage = ({ route }) => {
   const classes = useStyles();
-  let { shiftList } = useSelector((state) => state.shifts);
+  const dispatch = useDispatch();
   const currentUser = useSelector((state) => state.profile);
-  shiftList = datestringToDate(shiftList);
+  const shiftList = useSelector((state) => state.shifts.shiftList);
+  const currentShift = useSelector((state) => state.shift.currentShift);
+  const [events, setEvents] = React.useState([]);
 
-  if (!shiftList) {
-    return <CircularProgress />;
+  if (shiftList) {
+    if (!events) {
+      setEvents(datestringToDate(shiftList));
+    } else if (events.length !== shiftList.length) {
+      setEvents(datestringToDate(shiftList));
+    } else if (currentShift) {
+      if (Object.keys(currentShift).length > 0) {
+        if (
+          events.find((shift) => shift.id === currentShift.id).shift_users
+            .length !== currentShift.shift_users.length
+        ) {
+          setEvents(datestringToDate(shiftList));
+        }
+      }
+    }
   }
 
+  useEffect(() => {
+    if (!currentShift.id && route) {
+      const shift = shiftList.find((s) => s.id === parseInt(route.id, 10));
+      if (shift)
+        dispatch({
+          type: "skyportal/CURRENT_SHIFT",
+          data: shift,
+        });
+    } else if (currentShift) {
+      const updatedShift = shiftList.find((s) => s.id === currentShift.id);
+      // check if the shift shift_users length is different from the current shift
+      if (
+        updatedShift &&
+        updatedShift.shift_users.length !== currentShift.shift_users.length
+      ) {
+        dispatch({ type: "skyportal/CURRENT_SHIFT", data: updatedShift });
+      } else if (updatedShift) {
+        if (
+          Object.keys(updatedShift).length > 0 &&
+          Object.keys(currentShift).length > 0
+        ) {
+          if (
+            updatedShift.shift_users.some((user) =>
+              currentShift.shift_users.some(
+                (cUser) => cUser.needs_replacement === user.needs_replacement
+              )
+            )
```

Comment:
I'm not sure what you are checking here, but `needs_replacement` is a boolean so there may be an easier way to check this is true without two `some` calls.